buildlog/buildlog_test.go

```diff
@@ -0,0 +1,162 @@
+package buildlog
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type writeCloser struct {
+	io.Writer
+}
+
+func (wc writeCloser) Close() error {
+	return nil
+}
+
+func write(tb testing.TB, filepath string, inputs []string) {
+	tb.Helper()
+
+	for _, input := range inputs {
+		file, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+		if err != nil {
+			tb.Fatal(err)
+		}
+
+		_, err = file.Write([]byte(input))
+		if err != nil {
+			tb.Fatal(err)
+		}
+		//time.Sleep(1 * time.Second)
+
+		err = file.Close()
+		if err != nil {
+			tb.Fatal(err)
+		}
+	}
+}
+
+func TestRun(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		testInputs := []string{
+			`{"t": 1580198209, "m": "test 1", "n": 0, "s": "main"}` + "\n",
+			`{"t": 1580198222, "m": "test 2", "n": 1, "s": "main"}` + "\n",
+		}
+		go write(t, tmpFile.Name(), testInputs)
+
+		parent, cancel := context.WithCancel(context.Background())
+		writer := bytes.NewBuffer(nil)
+		l := log{
+			file:   tmpFile,
+			writer: writer,
+			ctx:    parent,
+			cancel: cancel,
+		}
+
+		errChan := l.Run()
+		_ = errChan
+
+		time.Sleep(3 * time.Second)
+		l.Stop()
+
+		expected := []string{
+			"2020-01-28 16:56:49 +0900 JST: test 1\n",
+			"2020-01-28 16:57:02 +0900 JST: test 2\n",
+		}
+		assert.Equal(t, strings.Join(expected, ""), writer.String())
+	})
+
+}
+
+func TestStop(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		testInputs := []string{
+			`{"t": 1580198209, "m": "test 1", "n": 0, "s": "main"}` + "\n",
+			`{"t": 1580198222, "m": "test 2", "n": 1, "s": "main"}` + "\n",
+		}
+
+		testInputsNotWritten := []string{
+			`{test}` + "\n",
+		}
+
+		go write(t, tmpFile.Name(), testInputs)
+
+		parent, cancel := context.WithCancel(context.Background())
+		writer := bytes.NewBuffer(nil)
+		l := log{
+			file:   tmpFile,
+			writer: writer,
+			ctx:    parent,
+			cancel: cancel,
+		}
+
+		errChan := l.Run()
+		_ = errChan
+
+		time.Sleep(3 * time.Second)
+		l.Stop()
+
+		go write(t, tmpFile.Name(), testInputsNotWritten)
+
+		expected := []string{
+			"2020-01-28 16:56:49 +0900 JST: test 1\n",
+			"2020-01-28 16:57:02 +0900 JST: test 2\n",
+		}
+		assert.Equal(t, strings.Join(expected, ""), writer.String())
```

Comment:
It's hard to understand the difference between `TestRun` and `TestStop`, so it's a good idea to add a comment.

Reply:
I added description for TestStop.